fix(curriculum): specify that letters.forEach() call should be nested (#53367)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-functional-programming-by-building-a-spreadsheet/6434759f78ec812264ff8f34.md

# --description--

> In your callback, make two function calls. Call `createLabel()` and pass `number` as the argument. You should see some numbers appear in your spreadsheet.

Reviewer: I would revert this sentence back to the way it was because we are not really making two function calls here. They only need to call the `createLabel` function

Author: `createLabel(number)` and `letters.forEach(...)` are two function calls.

Reviewer: It could be better worded to this then

```suggestion
In your callback, you will need to make two function calls. Start by calling `createLabel()` and pass `number` as the argument. You should see some numbers appear in your spreadsheet.
```

> Then call the `.forEach()` method on your `letters` array. Pass an empty callback function which takes a `letter` parameter.

Reviewer: Here, we should specific that the answer needs to be nested inside the callback

```suggestion
Below your `createLabel` function call, call the `.forEach()` method on your `letters` array. Pass an empty callback function which takes a `letter` parameter.
```

Author: The "two function calls" part above should imply that forEach is nested.

Author: Again, I'm okay with your suggestion. But the  code you say meets the specification doesn't have two function calls in the callback, it has one within and one without. 
If you'd still want me to do it your way, sure. I don't mind.

Reviewer: leave can leave the original as is

> assert.match(code, /letters\.forEach\(\s*\(?\s*letter\s*\)?\s*=>\s*\{\s*\}/)

Reviewer 2: ```suggestion
assert.match(code, /letters\.forEach\(\s*(\(\s*letter\s*\)|letter)\s*=>\s*\{\s*\}/)
```
While we are here...

Author: Are there valid patterns the existing regex doesn't match, but this one does ? Just curious.

Reviewer 2: It's the other way around, current will match invalid code, ie.:
```javascript
letters.forEach(letter) => {}
```

Author: Got it 👍

```

> Your callback function should be nested in `range(1, 99)`'s forEach callback

Reviewer: In an effort to simplify this, we should explain more with code examples here like this

```suggestion
Your `letters.forEach()` callback function should be nested inside the `range(1, 99).forEach(number => {}` callback function.
```

```js

> assert.match(code, /letters\.forEach\(\s*\(?\s*letter\s*\)?\s*=>\s*\{\s*\}.*\s*\}\)/)

Reviewer 2: This checks only if there's _nesting_, a bit safer is making sure it's within the `range(1, 99).forEach(number => {})`
```suggestion
assert.match(code, /range\s*\(\s*1\s*,\s*99\s*\)\s*.forEach\s*\(\s*(\(\s*number\s*\)|number)\s*=>\s*\{\s*[^}]*letters\.forEach\(\s*(\(\s*letter\s*\)|letter)\s*=>\s*\{\s*\}\s*\)\s*\}\s*\)/)
```
